content/testing.md

```diff
@@ -0,0 +1,683 @@
+    if (DDP._allSubscriptionsReady()) {
+      clearInterval(poll);
+      done();
+    }
+  }, 200);
+};
+
+// Utility to allow throwing inside callback
+const catchAsync = (done, cb) => () => {
+  try {
+    cb();
+  } catch (e) {
+    done(e);
+  }
+};
+
+if (Meteor.isClient) {
+  describe('data available when routed', () => {
+    beforeEach(done => {
+      generateData()
+        .then(() => FlowRouter.go('/'))
+        .nodeify(done);
+    });
+
+    describe('when logged out', () => {
+      it('has all public lists at homepage', () => {
+        assert.equal(Lists.find().count(), 3);
+      });
+
+      it('renders the correct list when routed to', done => {
+        const list = Lists.findOne();
+        FlowRouter.go('Lists.show', { _id: list._id });
+
+        // Wait for the router change to take affect
+        Tracker.afterFlush(catchAsync(done, () => {
+          assert.equal($('.title-wrapper').html(), list.name);
+
+          // Wait for all subscriptions triggered by this route to complete
+          waitForSubscriptions(catchAsync(done, () => {
+            assert.equal(Todos.find({listId: list._id}).count(), 3);
+            done();
+          }));
+        }));
+      });
+    });
+  });
+}
+```
+
+Of note here:
+
+ - Before running, each test sets up the data it needs using the `generateData` helper (see [the section on creating integration test data](#creating-integration-test-data) for more detail) then goes to the homepage.
+
+ - Although Flow Router doesn't take a done callback, we can use `Tracker.afterFlush` to wait for all it's reactive consequences to occur.
+
+ - Here we wrote a little utility (which could be abstracted into a general package) to wait for all the subscriptions which are created by the route change (the `todos.inList` subscription in this case) to become ready before checking their data.
+
+ - The `catchAsync` utility is useful because otherwise we don't see the results of failures inside callbacks. XXX: perhaps this test should be rewritten with promises to avoid this?
+
+
+<h3 id="running-full-app-tests">Running full-app tests</h3>
+
+To run the [full-app tests](#test-modes) in our application, we run:
+
+```
+meteor test --full-app --driver-package avital:mocha
+```
+
+When we connect to the test instance in a browser, we want to render a testing UI rather than our app UI, so the `mocha-web-reporter` package will hide any UI of our application and overlay it with its own. However the app continues to behave as normal, so we are able to route around and check the correct data is loaded.
+
+<h3 id="creating-integration-test-data">Creating data</h3>
+
+To create test data in full-app test mode, it usually makes sense to create some special test methods which we can call from the client side. Similar to the way we cleared the database using a method in the `beforeEach` in the [test data](#test-data) section above, we can call that method before running our tests. 
```

Comment:
Right - I mean we should call that out explicitly. Since we have one of these sections in each testing mode, it's a good place to call out why it's different from the previous once.